### PR TITLE
Updated latex parser rule for left right with unary function

### LINKED ISSUE #145: (URGENT) LaTeX math formula parsing crash
Reported by @jonsneyers:

https://github.com/metanorma/metanorma/issues/290
```
  \\[out_k = \\frac{1}{s}\\left(k == 0 ? 1 : \\sqrt{2}\\right)\\sum_{n = 0}^{s - 1}{in_{n} \\cdot \\cos\\left( \\frac{\\pi k}{s}\\left( n + \\frac{1}{2} \\right) \\right)}\\]
```

@suleman-uzair can you please help fix this ASAP? Thanks.

### PR DESCRIPTION
**Pull request by suleman-uzair**
Updated **Parser** rule and **Transformation** method.
Added spec of example from #145 

closes #145

### CODE REVIEW

**Comment by suleman-uzair** [comment]
@ronaldtse The **Plurimath** output may include additional "mrow" tags to wrap certain elements such as "left", "right", and unary functions, even if they are not present in your provided output. These "mrow" tags should not affect the presentation or functionality of the equation.
Secondly, the example you provided contains two "mfrac" tags, one contains "mrow" tags and the other does not.
can you guide me if there is any specific scenario that I'm missing here for the "mrow" tags?

__Expected output from the issue__

```
>   <math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
>     <mstyle displaystyle="true">
>       <mi>o</mi>
>       <mi>u</mi>
>       <msub>
>         <mi>t</mi>
>         <mi>k</mi>
>       </mrow>
>       </msub>
>       <mo>=</mo>
>       <mfrac>
>         <mrow><mn>1</mn></mrow>
>         <mrow><mi>s</mi></mrow>
>       </mfrac>
>       <mo>(</mo>
>       <mrow>
>         <mi>k</mi>
>         <mo>=</mo>
>         <mo>=</mo>
>         <mn>0</mn>
>         <mi>?</mi>
>         <mn>1</mn>
>         <mo>:</mo>
>         <msqrt>
>           <mn>2</mn>
>         </msqrt>
>       </mrow>
>       <mo>)</mo>
>       <munderover>
>         <mo>&#x2211;</mo>
>         <mrow>
>           <mi>n</mi>
>           <mo>=</mo>
>           <mn>0</mn>
>         </mrow>
>         <mrow>
>           <mi>s</mi>
>           <mo>&#x2212;</mo>
>           <mn>1</mn>
>         </mrow>
>       </munderover>
>       <mrow>
>         <mi>i</mi>
>         <msub>
>           <mi>n</mi>
>           <mi>n</mi>
>         </msub>
>         <mo>&#x22c5;</mo>
>         <mi>cos</mi>
>         <mo>(</mo>
>         <mrow>
>           <mfrac>
>             <mrow>
>               <mi>&#x3c0;</mi>
>               <mi>k</mi>
>             </mrow>
>             <mi>s</mi>
>           </mfrac>
>           <mo>(</mo>
>           <mrow>
>             <mi>n</mi>
>             <mo>+</mo>
>             <mfrac>
>               <mn>1</mn>
>               <mn>2</mn>
>             </mfrac>
>           </mrow>
>           <mo>)</mo>
>           </mrow>
>         <mo>)</mo>
>       </mrow>
>     </mstyle>
>   </math>

```
__End of output__
__Plurimath output__

```
>   <math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
>     <mstyle displaystyle="true">
>       <mo>[</mo>
>       <mi>o</mi>
>       <mi>u</mi>
>       <msub>
>         <mi>t</mi>
>         <mi>k</mi>
>       </msub>
>       <mo>=</mo>
>       <mfrac>
>         <mn>1</mn>
>         <mi>s</mi>
>       </mfrac>
>       <mrow>
>         <mo>(</mo>
>         <mrow>
>           <mi>k</mi>
>           <mo>=</mo>
>           <mo>=</mo>
>           <mn>0</mn>
>           <mi>?</mi>
>           <mn>1</mn>
>           <mo>:</mo>
>           <msqrt>
>             <mn>2</mn>
>           </msqrt>
>         </mrow>
>         <mo>)</mo>
>       </mrow>
>       <munderover>
>         <mo>&#x2211;</mo>
>         <mrow>
>           <mi>n</mi>
>           <mo>=</mo>
>           <mn>0</mn>
>         </mrow>
>         <mrow>
>           <mi>s</mi>
>           <mo>&#x2212;</mo>
>           <mn>1</mn>
>         </mrow>
>       </munderover>
>       <mrow>
>         <mi>i</mi>
>         <msub>
>           <mi>n</mi>
>           <mi>n</mi>
>         </msub>
>         <mo>&#x22c5;</mo>
>         <mrow>
>           <mi>cos</mi>
>           <mrow>
>             <mo>(</mo>
>             <mrow>
>               <mfrac>
>                 <mrow>
>                   <mi>&#x3c0;</mi>
>                   <mi>k</mi>
>                 </mrow>
>                 <mi>s</mi>
>               </mfrac>
>               <mrow>
>                 <mo>(</mo>
>                 <mrow>
>                   <mi>n</mi>
>                   <mo>+</mo>
>                   <mfrac>
>                     <mn>1</mn>
>                     <mn>2</mn>
>                   </mfrac>
>                 </mrow>
>                 <mo>)</mo>
>               </mrow>
>             </mrow>
>             <mo>)</mo>
>           </mrow>
>         </mrow>
>       </mrow>
>       <mo>]</mo>
>     </mstyle>
>   </math>
```

__End of output__

The output you provided as the expected result differs slightly from the output generated by **Plurimath**, as noted below. The difference between the two is the presence of some additional "mrow" tags in the expected output, which do not affect the presentation but only the comparison of the output.

Let me know if I need to change this.

**Comment by ronaldtse** [comment]
> The Plurimath output may include additional "mrow" tags to wrap certain elements such as "left", "right", and unary functions, even if they are not present in your provided output. These "mrow" tags should not affect the presentation or functionality of the equation.

The `<mrow>` tags are important inside the `<mo>` because they group things together (before the `paren`. In this case, they group the `(` and `)`, so that the left and right parentheses can be shown with the same height. If you try the MathML in Safari, you can see how they differ in presentation. Without the `<mrow>`, the left and right parentheses will be of different height.

> Secondly, the example you provided contains two "mfrac" tags, one contains "mrow" tags and the other does not.
> can you guide me if there is any specific scenario that I'm missing here for the "mrow" tags?

I think if there is more than one element inside an argument of the `mfrac` tag as arguments, then `mrow` should be used to group them.

**Comment by ronaldtse** [comment]
@suleman-uzair please also try this page for MathML rendering comparisons (by directly entering MathML):
* https://www.mathjax.org/

**Comment by suleman-uzair** [comment]
**Plurimath** output:
Screenshot from 2023-05-02 14-13-27
[Issue's](https://github.com/plurimath/plurimath/issues/145#:~:text=The%20correct%20Presentation%20MathML%20output%20should%20be%20(%40suleman%2Duzair%20use%20this%20as%20spec)%3A) output:
Screenshot from 2023-05-02 14-15-26

@ronaldtse Although the **Plurimath** output and the **[issue's](https://github.com/plurimath/plurimath/issues/145#:~:text=The%20correct%20Presentation%20MathML%20output%20should%20be%20(%40suleman%2Duzair%20use%20this%20as%20spec)%3A)** output appear to be identical in their presentation, there is a difference in the output code.

The **Plurimath** output includes a few additional "_mrow_" tags, which is due to the structure of how the data is stored.
Each "**Formula**" in our structure wraps its content in a "_mrow_" tag.
Let me know if this is correct or if it requires any further changes.

**Comment by ronaldtse** [comment]
@suleman-uzair where are you getting this rendering from? Thanks.

**Comment by suleman-uzair** [comment]
> @suleman-uzair where are you getting this rendering from? Thanks.

I'm using [TutorialsPoint](https://www.tutorialspoint.com/online_mathml_editor.php).

**Comment by ronaldtse** [comment]
@suleman-uzair is this ready to merge? Thanks.

File: spec/plurimath/asciimath_spec.rb [hunk 1]
```diff
@@ -4018,5 +4016,186 @@
...
+        expect(formula.to_mathml).to be_equivalent_to(mathml)
+        expect(formula.to_asciimath).to eql(asciimath)
+      end
+    end
+
+    context "contains example #74" do
+      let(:string) { 'ceil(x)' }
+
+      it 'returns parsed Asciimath to Formula' do
+        latex = '{\lceil x \rceil}'
```
Comment (ronaldtse) [inline spec/plurimath/asciimath_spec.rb:4072]: This is LaTeX, why is it saying Asciimath?

File: spec/plurimath/asciimath_spec.rb [hunk 2]
```diff
@@ -4018,5 +4016,186 @@
...
+        expect(formula.to_latex).to eql(latex)
+        expect(formula.to_mathml).to be_equivalent_to(mathml)
+        expect(formula.to_asciimath).to eql(asciimath)
+      end
+    end
+
+    context "contains example #76" do
+      let(:string) { '{x_i | i in I}' }
+
+      it 'returns parsed Asciimath to Formula' do
```
Comment (ronaldtse) [inline spec/plurimath/asciimath_spec.rb:4116]: LaTeX

File: spec/plurimath/asciimath_spec.rb [hunk 3]
```diff
@@ -4018,5 +4016,186 @@
...
+        expect(formula.to_latex).to eql(latex)
+        expect(formula.to_mathml).to be_equivalent_to(mathml)
+        expect(formula.to_asciimath).to eql(asciimath)
+      end
+    end
+
+    context "contains example #77" do
+      let(:string) { '"H"_nu^((1))(z)' }
+
+      it 'returns parsed Asciimath to Formula' do
```
Comment (ronaldtse) [inline spec/plurimath/asciimath_spec.rb:4146]: LaTeX?

File: spec/plurimath/asciimath_spec.rb [hunk 4]
```diff
@@ -4018,5 +4016,186 @@
...
+        MATHML
+        expect(formula.to_latex).to eql(latex)
+        expect(formula.to_mathml).to be_equivalent_to(mathml)
+        expect(formula.to_asciimath).to eql(asciimath)
+      end
+    end
+
+    context "contains example #78" do
+      let(:string) { '(a)_n' }
+
```
Comment (ronaldtse) [inline spec/plurimath/asciimath_spec.rb:4177]: LaTeX?

**Review by ronaldtse — APPROVED** [review]

**Comment by suleman-uzair** [comment]
@ronaldtse
The changes made to this file were pushed in [**PR #149**](https://github.com/plurimath/plurimath/pull/149), but they will be removed after the rebase. Separating the blocks for each conversion match is a possibility, but it will take some time since the changes need to be implemented in multiple files. As a result, I plan to create a new PR that includes these changes.

**Comment by suleman-uzair** [comment]
> OK. I also have a question -- are we supporting `\left( ... \right)`? See this for usage:
> 
> * https://tex.stackexchange.com/questions/12773/or-left-parentheses
> 
> All these following combinations are possible:
> 
> * `\left( ... \right)`
> * `\left ( ... \right )`
> * `\left< ... \right>`
> * etc.

Although we already support `\left( ... \right)`, the current list(mentioned below) of supported delimiters is limited.
1. `\left \{ \right \}`
2. `\left [ \right ]`
3. `\left ( \right )`
4. `\left . \right .`
Considering the vast number of delimiters that can be used with `\left( ... \right)`, I have found a [list](http://latexref.xyz/Delimiters.html) of commonly used delimiters to expand our support. Should we add these in our constants?

**Comment by ronaldtse** [comment]
> I have found a [list](http://latexref.xyz/Delimiters.html) of commonly used delimiters to expand our support. Should we add these in our constants?

@suleman-uzair great find here. Yes, let's do that. Thanks!

**Comment by ronaldtse** [comment]
Merging first.

**Pull request merged**